### PR TITLE
Convert Camerata to a gem, add a commandline tool

### PR DESCRIPTION
**Pull request by orangewolf**
Wrap the docker commands from each application in to a capable tool, saving us both key strokes and the need to maintain seperate files in multiple locations. Wrap the bin scripts so that they can be used from any directory and unwind the merge relases app merge release camerata double step to a single merge release of any given application while maintaining camerata's coordination abilities.

- `camerata up`
- `camerata cluster-ps yul-test main`

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,11 +1,16 @@
 # frozen_string_literal: true
 
 source 'https://rubygems.org'
-git_source(:github) { |repo| "https://github.com/#{repo}.git" }
+git_source(:github) { |_repo| 'https://github.com/#{repo}.git' }
 
+# Specify your gem's dependencies in camerata.gemspec
+gemspec
+
+# TODO: make these development dependencies?
```
Comment (fnibbit) [inline Gemfile:9]: Thinking about that question -- the question in my mind is, is there a use case where people will be wanting to use the camerata gem but not have the smoke tests available?

Comment (orangewolf) [reply Gemfile:9]: this makes a lot of sense to me.

File: camerata.gemspec [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  spec.authors       = ["Rob Kaufman"]
+  spec.email         = ["rob@notch8.com"]
+
+  spec.summary       = 'Coordinate services for YUL-DC project'
+  spec.description   = 'Command line tools and other combined services for the YUL-DC project. Shared Docker and deployment configuration, CI tools, etc live here.'
+  spec.homepage      = "https://github.com/yalelibrary/yul-dc-camerata"
+  spec.license       = "MIT"
+  spec.required_ruby_version = Gem::Requirement.new(">= 2.3.0")
+
+  spec.metadata["allowed_push_host"] = "TODO: Set to 'http://mygemserver.com'"
```
Comment (maxkadel) [inline camerata.gemspec:16]: Do you want to update this now, or later?

File: lib/camerata.rb [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+      run_with_exit("#{docker_compose} exec #{service} bundle exec rails console #{args.join(' ')}")
+    end
+    map rc: :console
+
+    desc "smoke ARGS", "Run the smoke tests against a running stack"
+    def smoke(*args)
+      run_with_exit("rspec #{smoke_path} #{args.join(' ')}")
+    end
+
+    desc "release APP VERSION", "Set a new version string for release of an application. For example `cam release blacklight 2.5.1`"
```
Comment (maxkadel) [inline lib/camerata.rb:129]: This makes me very happy

File: spec/camerata_cli_spec.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    it 'still shows method missing message for items not in shell scripts' do
+      expect { capture(:stdout) { cli.send('cluster-aa.sh') } }.to raise_error(NoMethodError)
+    end
+  end
+
+  xcontext 'up' do
+    let(:output) { capture(:stdout) { cli.up } }
+
+    it 'creates a docker compose file on the fly' do
+      expect(output).to match('ahaahhaha')
```
Comment (maxkadel) [inline spec/camerata_cli_spec.rb:51]: lol. I've joked with Bess that tests are to developers what marginalia were to medieval illuminators / scribes (they'd write little verses about how much their hands ached, or how chilly it was, or silly pictures of cats)

File: templates/management-compose.yml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+      POSTGRES_DB: management_yul_development
+      POSTGRES_HOST: db
+      POSTGRES_PASSWORD: password
+      POSTGRES_USER: postgres
+      SOLR_BASE_URL: http://solr:8983/solr
+      SOLR_CORE: blacklight-core
+      RAILS_RELATIVE_URL_ROOT: /management
+      RAILS_LOG_TO_STDOUT: 'true'
+    ports:
+      - "3001:3001"
```
Comment (maxkadel) [inline templates/management-compose.yml:16]: Thought for the future: It would probably help security if at some point we stop mapping the ports outside the containers in production-like environments.

File: README.md [hunk 6]
```diff
@@ -36,16 +82,18 @@ particular, we are making an effort to maintain a high degree of
...
+directory. One for the base, then a local override and a ecs override. These 
+files are composed together to create the compose file used for development or
+deploment as needed.
+
+file                              | contents
+--------------------------------- | ----------------------------------------------------------------------------------------------------
+`blacklight-compose.yml`          | compose definitions that are shared between all environments
+`blacklight-compose.local.yml`    | compose definitions required exclusively in a local docker environment
+`blacklight-compose.ecs.yml`      | compose definitions required for deployment to AWS ECS
+`.env`                            | No longer used. All env should be in Amazon SSM
```
Comment (maxkadel) [inline README.md:95]: Should we add any more breadcrumbs for the Amazon SSM here?

File: README.md [hunk 7]
```diff
@@ -196,25 +245,29 @@ root of this repo.
...
-## Releasing a new Camerata version
+3. `cam smoke`
+
+## Releasing a new dependency version
+
+1. Follow the release process laid out in the application README through to completion
+2. Set the version variable to the new version with `cam release APP_NAME VERSION_NUMBER`
+3. Start the applications with the new version and run the smoke test
+4. Deploy the applications (see deployment above) Example:
+`cam deploy-main.sh yul-test`
```
Comment (maxkadel) [inline README.md:267]: take off `.sh` here?

File: Gemfile [hunk 8]
```diff
@@ -1,11 +1,16 @@
 # frozen_string_literal: true
 
 source 'https://rubygems.org'
-git_source(:github) { |repo| "https://github.com/#{repo}.git" }
+git_source(:github) { |_repo| 'https://github.com/#{repo}.git' }
```
[the remark this answers is not in the document]
Comment (orangewolf) [reply Gemfile:4]: yep, that that why rubocop underscored the variable.

File: camerata.gemspec (hunk 2 above)
Comment (orangewolf) [reply camerata.gemspec:16]: right now it s"no where" so this is an ok setting

File: templates/management-compose.yml (hunk 5 above)
Comment (orangewolf) [reply templates/management-compose.yml:16]: small edits:

Thought for the future: It would probably help ~~security~~ everything if at some point we stop mapping the ports outside the containers ~~in production-like environments~~ everywhere.

Comment (orangewolf) [reply templates/management-compose.yml:16]: ;-)

File: README.md (hunk 7 above)
Comment (orangewolf) [reply README.md:267]: yep, missed that one. it was in my clipboard when I fixed the others.  One thing I did for the transition is that camerata allows the .sh versions to make it easier for folks who already remember them.

File: lib/camerata.rb (hunk 3 above)
Comment (orangewolf) [reply lib/camerata.rb:129]: =-D

File: README.md (hunk 6 above)
Comment (orangewolf) [reply README.md:95]: I agree. there are a set of tickets in the next sprint about SSM.  I left some low hanging fruit here on purpose as I want others to get used to making edits and I really tried to keep SSM out of scope for this work but ultimatly it was going to be 4 ours to use it and 16 hours to not use it. So I think the breadcrumbs are needed but out of scope here.

File: .rubocop.yml [hunk 9]
```diff
@@ -1,13 +1,18 @@
...
+Metrics/BlockLength:
+  Max: 38
   Exclude:
-    - spec/deploy_spec.rb
+    - 'spec/**/*.rb'
+    - 'smoke_spec/**/*.rb'
 
 RSpec/ExampleLength:
-  Enabled: false
+  Exclude:
```
Comment (fnibbit) [inline .rubocop.yml:15]: Out of curiosity: why the change from disabling RSpec/ExampleLength to excluding all the specs from it?  I think both wind up doing the same thing.

File: README.md (hunk 6 above)
Comment (fnibbit) [reply README.md:95]: Ok, can you tell me them though?

File: lib/camerata.rb [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+    def stop(*args)
+      ensure_env
+      run("#{docker_compose} stop #{args.join(' ')}")
+      run_with_exit("rm -rf tmp/pids/*")
+    end
+
+    desc "restart", "restarts the specified running service, defaults to all"
+    def restart(*args)
+      ensure_env
+      run_with_exti("#{docker_compose} restart #{args.join(' ')}")
```
Comment (fnibbit) [inline lib/camerata.rb:45]: "run_with_exit" yeah?

File: .rubocop.yml (hunk 9 above)
Comment (orangewolf) [reply .rubocop.yml:15]: the `Enabled: false` was still triggering on the smoke specs for some reason. this fixed it so I didn't spend more time trying to figure out why.

File: lib/camerata.rb (hunk 10 above)
Comment (orangewolf) [reply lib/camerata.rb]: totally. can you guess which method I added very very last after I'd done most of my testing? thank you for catching that

File: README.md [hunk 11]
```diff
@@ -8,14 +8,52 @@ Coordinate services for YUL-DC project
...
+
+All buildin commands can be listed with `cam help` and individual usage 
+information is available with `cam help COMMAND`.  Please note that deployment 
+commands (found in the `./bin` directory) are pass through and are therefor not 
+listed by the help command.  See th usage for those below. 
+
+To start the application stack, run `cam up`. This is the equivalent of running 
+`docker-compose up blacklight`. This starts all of the applications as they are 
+all dependencies of yul-blacklight. Camerata is smart. If you start `cam up` from 
+a blacklight code check out it will mount that code for local development
```
Comment (JzGo) [inline README.md:48]: Does this effectively mean I should be able to `cam up` from the blacklight repo? Wasn't able to do that but I could be off in my interpretation here. I ask bc I know it's a goal to reduce the number of ways we `up` our apps in development and thought this was where we were going to do that.

Comment (JzGo) [reply README.md:48]: We did mention a possible teach out for this shift. Perhaps that is the place for this question.
